src/machma/db.py

```diff
@@ -0,0 +1,125 @@
+
+import contextlib
+import functools
+import logging
+
+import nr.proxy
+from sqlalchemy import create_engine, Column, ForeignKey, Integer, String
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import relationship, sessionmaker
+from sqlalchemy import func as F
+
+from .utils.sqlalchemy.guc import GucHelper
+
+LOGGER = logging.getLogger(__name__)
+Base = declarative_base()
+Session = sessionmaker()
+session = nr.proxy.threadlocal[Session](
+    name=__name__ + '.session',
+    error_message=
+        '({name}) No SqlAlchemy session is available. Ensure that you are using the '
+        'make_session() context manager before accessing the global session proxy.',
+)
+
+__all__ = [
+    'Base',
+    'Session',
+    'session',
+    'initialize_db',
+    'Exercise',
+    'ExerciseAlias',
+    'User',
+    'UserReps',
+    'F',
+]
+
+
+def initialize_db(*args, create_tables: bool = False, **kwargs):
+    """
+    Creates an SqlAlchemy engine and configures the #Session class. The arguments are
+    forwarded to the #create_engine() function (see [1]).
+
+    [1]: https://docs.sqlalchemy.org/en/13/core/engines.html#sqlalchemy.create_engine
+    """
+
+    LOGGER.info('Initializing SqlAlchemy Session')
+    engine = create_engine(*args, **kwargs)
+    Session.configure(bind=engine)
+
+    if create_tables:
+        Base.metadata.create_all(engine)
+
+
+@contextlib.contextmanager
+def make_session() -> None:
+    """
+    A context manager that creates a new #Session object and makes it available in the global
+    #session proxy object.
+    """
+
+    nr.proxy.push(session, Session())
+    try:
+        yield
+    except:
+        session.rollback()
+        raise
+    else:
+        session.commit()
+    finally:
+        nr.proxy.pop(session)
+
+
+def async_session(func):
+    """
+    Decorator for an async function that wraps it in a #make_session() call to ensure
+    that a session is available.
+    """
+
+    @functools.wraps(func)
+    async def wrapper(*args, **kwargs):
+        with make_session():
+            return await func(*args, **kwargs)
+
+    return wrapper
+
+
+def get(entity, **pks) -> GucHelper:
```

Comment:
The name is confusing. Especially since this leads one to use `session.query` for some queries and `db.get` for other queries.

Reply:
I used this name because I think it reads nicely, i.e. `get then_update or_create`.  We could introduce an alias for `session.query` as `db.query` (it may be less confusing to have `db.query` and `db.get` instead of `session.query` and `db.get`).

I don't think there's a way to write the same starting with `session.query()`.

Feel free to suggest an alternative name for the function.